orquesta/conducting.py

```diff
@@ -730,6 +761,50 @@ def update_task_state(self, task_id, route, event):
             state_ctx = {'__state': self.workflow_state.serialize()}
             current_ctx = dict_util.merge_dicts(current_ctx, state_ctx, True)
 
+            # This decides whether this task would be retried or not when 'retry' is specified
+            may_retry = (
+                hasattr(task_spec, 'retry') and
+                hasattr(task_spec.retry, 'when') and
+                task_spec.has_retry()
+            )
+            if may_retry:
+                spec_retry_count = expr_base.evaluate(task_spec.retry.count, current_ctx)
+
+                # This is a measure to pass through this processing safely when invalid values
+                # are specified in retry parameter
+                try:
+                    # Current task will be retried when all these conditions are satisfied
+                    #   - new status is same as the one which is specified one.
+                    #   - task has never been repeated for specified times.
+                    will_retry = (
+                        int(spec_retry_count) > task_state_entry['ctxs']['retry']['count'] and
+                        expr_base.evaluate(task_spec.retry.when, current_ctx)
+                    )
+                    if will_retry:
+                        # This turns task_status back to old one.
+                        task_state_entry['status'] = new_task_status = old_task_status
+
+                        # Increment task count to keep track of how many times task is repeated.
+                        task_state_entry['ctxs']['retry']['count'] += 1
+
+                        # This sets timer when this task would be run when valid delay parameter is
+                        if task_spec.retry.delay:
+                            task_state_entry['ctxs']['retry']['running_time'] = (
+                                int(datetime.datetime.now().strftime('%s')) +
+                                int(expr_base.evaluate(task_spec.retry.delay, current_ctx))
+                            )
+
+                        # This task is pushed back into the task queue again.
+                        self.workflow_state.add_staged_task(
+                            task_id,
+                            route,
+                            ctxs=task_state_entry['ctxs']['in'],
+                            prev=task_state_entry['prev'],
```

Comment:
Should probably include the `delay` here so it can be written to the staged task, which will be passed to StackStorm for processing.